boa/vm/fork.py

```diff
@@ -37,29 +37,38 @@ class CachingRPC(RPC):
     def __init__(self, rpc: RPC, cache_file: str = DEFAULT_CACHE_DIR):
         # (default to memory db plyvel not found or cache_file is None)
         self._rpc = rpc
-        self._init_mem_db()
-        if cache_file is not None:
-            try:
-                from boa.util.leveldb import LevelDB
-
-                print("(using leveldb)", file=sys.stderr)
-
-                cache_file = os.path.expanduser(cache_file)
-                # use CacheDB as an additional layer over disk
-                # (ideally would use leveldb lru cache but it's not configurable
-                # via LevelDB API).
-                self._db = CacheDB(LevelDB(cache_file), cache_size=1024 * 1024)  # type: ignore
-            except ImportError:
-                # plyvel not found
-                pass
+        #Only attempt DB initialization if its not done prior
+        #It appears __init__ is being called multiple times, which causes LOCK file error from levelDB.
+        try: self._db
+        except AttributeError:
+            self._init_mem_db()
+            if cache_file is not None:
+                try:
+                    from boa.util.leveldb import LevelDB
+
+                    print("(using leveldb)", file=sys.stderr)
+
+                    cache_file = os.path.expanduser(cache_file)
+                    # use CacheDB as an additional layer over disk
+                    # (ideally would use leveldb lru cache but it's not configurable
+                    # via LevelDB API).
+                    self._db = CacheDB(LevelDB(cache_file), cache_size=50 * 1024 * 1024)  # type: ignore
+                except ImportError:
+                    # plyvel not found
+                    pass
 
     # _loaded is a cache for the constructor.
     # reduces fork time after the first fork.
     _loaded: dict[tuple[str, str], "CachingRPC"] = {}
     _pid: int = os.getpid()  # so we can detect if our fds are bad
 
     def _init_mem_db(self):
-        self._db = MemoryDB(lrudict(1024 * 1024))
+        #Only attempt DB initialization if its not done prior
+        #It appears _init_mem_db is being called multiple times, which causes
```

Comment:
This is called here
https://github.com/vyperlang/titanoboa/blob/d50f0bc53855b314cdb9587a6e4d29f438782bf4/boa/vm/py_evm.py#L399

I think that is the right behavior as you may want to fork a different rpc.
However, the line I mentioned is maybe ignoring the cache file (@charles-cooper please confirm)

Reply:
`_init_mem_db` or `__init__` is being called on each request, firstly overriding the levelDB cache, then replacing `MemoryDB` with a fresh empty one on each call...

Reply:
maybe we should rename `_init_mem_db()` to `_init_db()` and have it do the choosing whether to use MemoryDB or LevelDB